libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3586,12 +3587,59 @@ void TypeChecker::endVisit(Literal const& _literal)
 
 void TypeChecker::endVisit(UsingForDirective const& _usingFor)
 {
-	if (m_currentContract->isInterface())
-		m_errorReporter.typeError(
-			9088_error,
-			_usingFor.location(),
-			"The \"using for\" directive is not allowed inside interfaces."
+	Type const* normalizedType = nullptr;
+	if (TypeName const* typeName = _usingFor.typeName())
+	{
+		solAssert(typeName->annotation().type);
+		normalizedType = TypeProvider::withLocationIfReference(DataLocation::Storage, typeName->annotation().type);
+		solAssert(normalizedType);
+	}
+
+	for (ASTPointer<IdentifierPath> const& path: _usingFor.functionsOrLibrary())
+	{
+		solAssert(path->annotation().referencedDeclaration);
+		// No type checking if it is a library.
+		FunctionDefinition const* functionDefinition =
+			dynamic_cast<FunctionDefinition const*>(path->annotation().referencedDeclaration);
+		if (!functionDefinition)
+			continue;
```

Comment:
I'd consider to instead do something like
```
if (!_usingFor.usesBraces())
{
  // assert that _usingFor.functionsOrLibrary() is a single library
  return;
}
```
Before the loop and then assert that ``functionDefinition`` is not ``nullptr`` here.

But also fine as is.

Reply:
Restructured it a bit.